ludwig/decoders/sequence_decoders.py

```diff
@@ -13,18 +13,227 @@
+        super().__init__()
+        self.hidden_size = hidden_size
+        self.vocab_size = vocab_size
+        self.embedding = nn.Embedding(vocab_size, hidden_size)
+        self.relu = nn.ReLU()
+        if cell_type == "gru":
+            self.rnn = nn.GRU(hidden_size, hidden_size, batch_first=True)
+        else:
+            self.rnn = nn.RNN(hidden_size, hidden_size, batch_first=True)
+        self.out = nn.Linear(hidden_size, vocab_size)
+
+    def forward(self, input: torch.Tensor, hidden: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Runs a single decoding time step.
+
+        Modeled off of https://pytorch.org/tutorials/intermediate/seq2seq_translation_tutorial.html.
+
+        Args:
+            input: [batch_size] tensor with the previous step's predicted symbol.
+            hidden: [batch_size, hidden_size] tensor with the previous step's hidden state.
+
+        Returns:
+            Tuple of two tensors:
+            - output: [batch_size, vocab_size] tensor with the logits.
+            - hidden: [batch_size, hidden_size] tensor with the hidden state for the next time step.
+        """
+        # Unsqueeze predicted tokens.
+        input = input.unsqueeze(1).to(torch.int)
+        output = self.relu(self.embedding(input))
+        output, hidden = self.rnn(output, hidden)
+        output_logits = self.out(output)
+        return output_logits, hidden
+
+
+class LSTMDecoder(nn.Module):
+    """LSTM-based decoder."""
+
+    def __init__(self, hidden_size: int, vocab_size: int):
+        super().__init__()
+        self.hidden_size = hidden_size
+        self.vocab_size = vocab_size
+        self.embedding = nn.Embedding(vocab_size, hidden_size)
+        self.relu = nn.ReLU()
+        self.lstm = nn.LSTM(hidden_size, hidden_size, batch_first=True)
+        self.out = nn.Linear(hidden_size, vocab_size)
+
+    def forward(
+        self, input: torch.Tensor, hidden_state: torch.Tensor, cell_state: torch.Tensor
+    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        """Runs a single decoding time step.
+
+        Modeled off of https://pytorch.org/tutorials/intermediate/seq2seq_translation_tutorial.html.
+
+        Args:
+            input: [batch_size] tensor with the previous step's predicted symbol.
+            hidden_state: [batch_size, hidden_size] tensor with the previous step's hidden state.
+            cell_state: [batch_size, hidden_size] tensor with the previous step's cell state.
+
+        Returns:
+            Tuple of 3 tensors:
+            - output: [batch_size, vocab_size] tensor with the logits.
+            - hidden_state: [batch_size, hidden_size] tensor with the hidden state for the next time step.
+            - cell_state: [batch_size, hidden_size] tensor with the cell state for the next time step.
+        """
+        # Unsqueeze predicted tokens.
+        input = input.unsqueeze(1).to(torch.int)
+        output = self.relu(self.embedding(input))
```

Comment:
why rely?

Reply:
Relu was used in the [reference implementation](https://pytorch.org/tutorials/intermediate/seq2seq_translation_tutorial.html#simple-decoder).

I'm leaning towards keeping it since it matches the reference implementation, but also happy to take it out if there's a good reason to remove it.

The model seems to train about the same either way.

Reply:
Discussed offline. Decided to remove the relu.